File: matlab/run_aucpr_from_precomputed.m
function run_aucpr_from_precomputed(net_file, perturbed_net_file, pairs_file, varargin)
% Run CH link prediction + AUCPR evaluation from pre-computed .mat files.
%
% Usage (default paths):
%   run_aucpr_from_precomputed()
%
% Usage (custom paths):
%   run_aucpr_from_precomputed('../matrix/Yeast_DIP_net.mat', ...
%       '../matrix/network_perturbed_10percent_GSP_GSN_Yeast_noconn.mat', ...
%       '../matrix/list_pairs_10percent_GSP_GSN_Yeast_noconn.mat')

if nargin < 1 || isempty(net_file)
    net_file = '../matrix/Yeast_DIP_net.mat';
end
if nargin < 2 || isempty(perturbed_net_file)
    perturbed_net_file = '../matrix/network_perturbed_10percent_GSP_GSN_Yeast_noconn.mat';
end
if nargin < 3 || isempty(pairs_file)
    pairs_file = '../matrix/list_pairs_10percent_GSP_GSN_Yeast_noconn.mat';
end

ip = inputParser;
addParameter(ip, 'output_file', '../table/table_auc_pr_CH_Yeast_DIP_net.xlsx');
parse(ip, varargin{:});
filename = ip.Results.output_file;

% ---- Load data ----
fprintf('Loading %s\n', net_file);
load(net_file, 'x_lcc');

fprintf('Loading %s\n', perturbed_net_file);
load(perturbed_net_file, 'L_net');

fprintf('Loading %s\n', pairs_file);
load(pairs_file, 'L_pairs');

num_sims = length(L_net);
fprintf('Loaded %d simulations.\n\n', num_sims);

% ---- CH link prediction ----
methods = {'RA_L2','CH1_L2','CH2_L2','CH3_L2','CH3.1_L2','iLCL_L2', ...
           'RA_L3','CH1_L3','CH2_L3','CH3_L3','CH3.1_L3','iLCL_L3'};

S = cell(num_sims, 1);
for i = 1:num_sims
    fprintf('CH link prediction: simulation %d/%d\n', i, num_sims);
    S{i} = CHA_linkpred_monopartite_final(L_net{i}, methods);
end

results_dir = '../network_similarities/CH_L2_L3/results/';
if ~exist(results_dir, 'dir'), mkdir(results_dir); end
save([results_dir 'CH_L2_L3_scores_net_perturbed.mat'], 'S', '-v7.3');
fprintf('CH scores saved.\n\n');

% ---- AUCPR evaluation ----
CH_methods = {'RA_L2','CH1_L2','CH2_L2','CH3_L2','CH3_1_L2','iLCL_L2', ...
              'RA_L3','RA_L3_subranking','CH1_L3','CH2_L3','CH3_L3','CH3_1_L3','iLCL_L3'};

res_CH_AUCPR = zeros(num_sims, length(CH_methods));

for i = 1:num_sims
    S_tmp = table2array(S{i});
    CH_mat = zeros(size(x_lcc));
    ind = sub2ind(size(CH_mat), S_tmp(:,1), S_tmp(:,2));
    P = L_pairs{i,1};
    N = L_pairs{i,2};
    labels = repelem([1, 0], size(P,1))';

    for j = 1:length(CH_methods)
        CH_mat(ind) = S_tmp(:, j+2);
        CH_mat_sym = CH_mat + CH_mat.';
        pos_scores = arrayfun(@(k) CH_mat_sym(P(k,1), P(k,2)), 1:size(P,1))';
        neg_scores = arrayfun(@(k) CH_mat_sym(N(k,1), N(k,2)), 1:size(N,1))';
        measures = prediction_evaluation([pos_scores; neg_scores], labels);
        res_CH_AUCPR(i, j) = measures.auc_pr;
    end
    fprintf('AUCPR evaluation: simulation %d/%d done\n', i, num_sims);
end

% ---- Build and write output table ----
tp = round(res_CH_AUCPR', 3);
tr = zeros(size(tp));
for j = 1:num_sims
    tr(:,j) = tiedrank(-tp(:,j));
end
t = [mean(tr,2), mean(tp,2), tp];
[t, idx] = sortrows(t, [1, -2]);

mean_ranking    = t(:,1)';
mean_aucpr      = t(:,2)';
sim_data        = t(:,3:end)';
col_methods     = CH_methods(idx);
sim_names       = arrayfun(@(k) sprintf('Realization_%02d', k), (1:num_sims)', 'UniformOutput', false);

table_dir = fileparts(filename);
if ~isempty(table_dir) && ~exist(table_dir, 'dir'), mkdir(table_dir); end

writetable(cell2table([{'methods'},      col_methods]),            filename, 'Range', 'A1', 'WriteVariableNames', false);
writetable(cell2table([{'mean_ranking'}, num2cell(mean_ranking)]), filename, 'Range', 'A2', 'WriteVariableNames', false);
writetable(cell2table([{'mean_aucpr'},   num2cell(mean_aucpr)]),   filename, 'Range', 'A3', 'WriteVariableNames', false);
writetable(cell2table([sim_names,        num2cell(sim_data)]),     filename, 'Range', 'A4', 'WriteVariableNames', false);

fprintf('\nResults written to %s\n', filename);
disp(round(mean(res_CH_AUCPR), 3));
end

File: python/run_aucpr_from_precomputed.py
import argparse
import os
import pickle

import numpy as np
import pandas as pd
from scipy.io import loadmat
from scipy.stats import rankdata

from ch_linkpred import cha_linkpred_monopartite
from prediction_evaluation import prediction_evaluation

CH_METHODS = [
    'RA_L2', 'CH1_L2', 'CH2_L2', 'CH3_L2', 'iLCL_L2', 'CH3.1_L2',
    'RA_L3', 'RA_L3_subranking',
    'CH1_L3', 'CH2_L3', 'CH3_L3', 'iLCL_L3', 'CH3.1_L3',
]


def _load_mat(filepath):
    return loadmat(filepath)


def _load_field(data, candidates, label):
    for name in candidates:
        if name in data:
            return data[name]
    available = [k for k in data if not k.startswith('_')]
    raise KeyError(f"Variable not found in {label}. Available: {available}. Expected: {candidates}")


def _to_dense(mat):
    return mat.toarray() if hasattr(mat, 'toarray') else np.asarray(mat)


def run_aucpr_from_precomputed(
    net_file='../matrix/Yeast_DIP_net.mat',
    perturbed_file='../matrix/network_perturbed_10percent_GSP_GSN_Yeast_noconn.mat',
    pairs_file='../matrix/list_pairs_10percent_GSP_GSN_Yeast_noconn.mat',
    output_file='../table/table_auc_pr_CH_Yeast_DIP_net.xlsx',
    n_jobs=None,
):
    print(f"Loading network from {net_file}")
    net_data = _load_mat(net_file)
    x_lcc = _load_field(net_data, ['x_lcc', 'A', 'adj', 'network'], net_file)
    N = _to_dense(x_lcc).shape[0]
    print(f"  Network size: {N} nodes")

    print(f"Loading perturbed networks from {perturbed_file}")
    L_net = _load_field(_load_mat(perturbed_file), ['L_net', 'L_nets'], perturbed_file)

    print(f"Loading pairs from {pairs_file}")
    L_pairs = _load_field(_load_mat(pairs_file), ['L_pairs', 'pairs'], pairs_file)

    num_sims = L_net.shape[0]
    print(f"  {num_sims} simulations loaded\n")

    methods = ['RA_L2', 'CH1_L2', 'CH2_L2', 'CH3_L2', 'iLCL_L2', 'CH3.1_L2',
               'RA_L3', 'CH1_L3', 'CH2_L3', 'CH3_L3', 'iLCL_L3', 'CH3.1_L3']

    S_list = []
    for i in range(num_sims):
        print(f"CH link prediction: simulation {i+1}/{num_sims}")
        df = cha_linkpred_monopartite(_to_dense(L_net[i, 0]), methods, n_jobs=n_jobs)
        S_list.append(df)

    results_dir = '../network_similarities/CH_L2_L3/results/'
    os.makedirs(results_dir, exist_ok=True)
    with open(os.path.join(results_dir, 'CH_L2_L3_scores_net_perturbed.pkl'), 'wb') as f:
        pickle.dump(S_list, f)
    print(f"CH scores saved.\n")

    res_CH_AUCPR = np.zeros((num_sims, len(CH_METHODS)))
    for i in range(num_sims):
        df = S_list[i]
        node1_0 = df['node1'].values - 1
        node2_0 = df['node2'].values - 1
        P_pairs = np.asarray(L_pairs[i, 0], dtype=int)
        N_pairs = np.asarray(L_pairs[i, 1], dtype=int)
        n_pos = P_pairs.shape[0]
        labels = np.concatenate([np.ones(n_pos), np.zeros(n_pos)])

        for j, method in enumerate(CH_METHODS):
            ch_mat = np.zeros((N, N))
            ch_mat[node1_0, node2_0] = df[method].values
            ch_mat_sym = ch_mat + ch_mat.T
            pos_scores = ch_mat_sym[P_pairs[:, 0] - 1, P_pairs[:, 1] - 1]
            neg_scores = ch_mat_sym[N_pairs[:, 0] - 1, N_pairs[:, 1] - 1]
            measures = prediction_evaluation(np.concatenate([pos_scores, neg_scores]), labels)
            res_CH_AUCPR[i, j] = measures['auc_pr']

        print(f"AUCPR evaluation: simulation {i+1}/{num_sims} done")

    tp = np.round(res_CH_AUCPR.T, 3)
    tr = np.zeros_like(tp)
    for j in range(num_sims):
        tr[:, j] = rankdata(-tp[:, j], method='average')
    t = np.column_stack([tr.mean(axis=1), tp.mean(axis=1), tp])
    sort_idx = np.lexsort((-t[:, 1], t[:, 0]))
    t = t[sort_idx]

    col_methods_sorted = [CH_METHODS[i] for i in sort_idx]
    sim_data = t[:, 2:].T
    sim_names = [f'Realization_{k+1:02d}' for k in range(num_sims)]

    os.makedirs(os.path.dirname(os.path.abspath(output_file)), exist_ok=True)
    rows = (
        [['methods']      + col_methods_sorted] +
        [['mean_ranking'] + list(np.round(t[:, 0], 3))] +
        [['mean_aucpr']   + list(np.round(t[:, 1], 3))] +
        [[name] + list(np.round(row, 3)) for name, row in zip(sim_names, sim_data)]
    )
    with pd.ExcelWriter(output_file, engine='openpyxl') as writer:
        pd.DataFrame(rows).to_excel(writer, index=False, header=False)

    print(f"\nResults written to {output_file}")
    print(np.round(res_CH_AUCPR.mean(axis=0), 3))


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--net_file',       default='../matrix/Yeast_DIP_net.mat')
    parser.add_argument('--perturbed_file', default='../matrix/network_perturbed_10percent_GSP_GSN_Yeast_noconn.mat')
    parser.add_argument('--pairs_file',     default='../matrix/list_pairs_10percent_GSP_GSN_Yeast_noconn.mat')
    parser.add_argument('--output_file',    default='../table/table_auc_pr_CH_Yeast_DIP_net.xlsx')
    parser.add_argument('--n_jobs',         type=int, default=None, help='Number of parallel workers (1=serial, None=all CPUs)')
    args = parser.parse_args()
    run_aucpr_from_precomputed(args.net_file, args.perturbed_file, args.pairs_file, args.output_file, args.n_jobs)
